sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/SdkRequestContext.java

```diff
@@ -0,0 +1,135 @@
+ * configure request metadata using {@link RequestOptions}.
+ * <p>
+ * This class does not allow any modifications to the request metadata provided by the user.
+ */
+public final class SdkRequestContext extends RequestOptions {
+    // TODO (limolkova) do we need to make it public and probably move to another class?
+    private static final String PROGRESS_REPORTER_KEY = "progressReporter";
+    static final SdkRequestContext NONE = new SdkRequestContext(null, null);
+    private static final ClientLogger LOGGER = new ClientLogger(SdkRequestContext.class);
+    private final InstrumentationContext childInstrumentationContext;
+
+    private SdkRequestContext(RequestOptions options, InstrumentationContext instrumentationContext) {
+        super(options);
+
+        this.childInstrumentationContext = instrumentationContext;
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options and instrumentation context.
+     *
+     * @param options The {@link RequestOptions} to be used.
+     * @param instrumentationContext The {@link InstrumentationContext} to be used.
+     * @return A new instance of {@link SdkRequestContext} or, when optimization is possible, the existing instance of {@link SdkRequestContext}
+     * provided in the {@code options} parameter.
+     */
+    public static SdkRequestContext create(RequestOptions options, InstrumentationContext instrumentationContext) {
+        if (options instanceof SdkRequestContext
+            && (instrumentationContext == null || !instrumentationContext.isValid())) {
+            return (SdkRequestContext) options;
+        }
+
+        return new SdkRequestContext(options, instrumentationContext);
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options.
+     * @param options The {@link RequestOptions} to be used.
+     * @return A new instance of {@link SdkRequestContext} or, when optimization is possible, the existing instance of {@link SdkRequestContext}
+     * provided in the {@code options} parameter.
+     */
+    public static SdkRequestContext create(RequestOptions options) {
+        if (options instanceof SdkRequestContext) {
+            return (SdkRequestContext) options;
+        }
+
+        return new SdkRequestContext(options, null);
+    }
+
+    @Override
+    public InstrumentationContext getInstrumentationContext() {
+        if (childInstrumentationContext != null && childInstrumentationContext.isValid()) {
+            return childInstrumentationContext;
+        }
+
+        return super.getInstrumentationContext();
+    }
+
+    /**
+     * Get the {@link ProgressReporter} from the request context.
+     *
+     * @return The {@link ProgressReporter} if present, otherwise null.
+     */
+    public ProgressReporter getProgressReporter() {
+        Object progressReporter = super.getData(PROGRESS_REPORTER_KEY);
+        if (progressReporter instanceof ProgressReporter) {
```

Comment:
I'm happy to see convenience on getting `ProgressReporter` from `Context` but I feel uncertain on whether I'd prefer this to be an explicit field on this class.

Right now, we have a possible pitfall in capability as a customer could pass `ProgressReporter` to Storage's parallel upload but there's no way to manage the sub-`ProgressReporter` for each individual upload call. I also believe this was solved for tracing with the factory method accepting `InstrumentationContext` but not resolved for progress reporting.

Reply:
I've added a factory method that takes progress reporter.